Makes upgrading medical sleepers do something, adds emag effects.

mikomyazaki opened the pull request:
Currently some medical machines are upgradeable, but this has no effect on their operation. I've done some sleeper upgrades so far. Depending on feedback/interest I will continue to other medical machines.

- Upgrading the Sleeper will add Kelotane (with nano-manipulators) and Hyrolanin (with pico-manipulators). These are quite basic chemicals on purpose.
- Upgrading the scanning module will decrease synthesis power requirements (down to 1/2 then 1/3) and increase pumping/filtering speed up to 33/66% faster.
- Emagging the sleeper will allow it to produce hair remover and chloral. There is a warning on examine that it is able to, and anyone could see them on the list.

So far as I know none of this allows production of further chemicals in the patient's bloodstream through the sleeper alone. 

I'd welcome feedback from medical players in #feedback about how powerful these upgrades would be. Don't want it to be too over the top, just make a useful difference. I'm a bit ambivalent about chloral through emagging, since the machine can already produce a sedative.

comma commented on code/game/machinery/Sleeper.dm:
```diff
@@ -278,10 +286,46 @@
+	for (var/obj/item/weapon/stock_parts/manipulator/M in component_parts) // manipulators unlock new drugs
+		T += M.rating
+
+	if (T >= 4 && !(upgrade_chemicals in available_chemicals))
```
can skip `!(upgrade_chemicals in available_chemicals)` by doing `available_chemicals |= upgrade_chemicals` instead of +=, it'll not add duplicates

comma replied:
I think it would be cleaner to move base chems into separate list and makethis whole block like
```
available_chemicals = base_chemicals
if(T >= 4)
  available_chemicals += upgrade_chemicals
if(T >= 6)
  available_chemicals += upgrade2_chemicals
```

Melioa commented:
Wouldn't someone know that the Sleeper's been emagged with the new chemicals available, regardless of medicine skill?

mikomyazaki commented:
> Wouldn't someone know that the Sleeper's been emagged with the new chemicals available, regardless of medicine skill?

Yes. I will just make it possible to see the Examine message with medical skill. The chemicals will show for everyone -- But there is no reason to know they are not always available if you have untrained/basic medical skill.

mikomyazaki commented:
Decided I can't think of good stuff for the medical scanners at the moment. So I'll be changing this PR to be just about the sleepers - So it is ready for review.

afterthought2 commented on code/game/machinery/Sleeper.dm:
```diff
@@ -41,6 +47,9 @@
 		if(occupant)
 			occupant.examine(user)
+	if (emagged && user.skill_check(SKILL_MEDICAL, SKILL_EXPERT))
+		to_chat(usr, "The sleeper chemical synthesis controls look tampered with.")
```
`user`, not `usr`. Also this needs to be behind the prior block's check (`. && Adjacent`).

afterthought2 commented on code/game/machinery/Sleeper.dm:
```diff
@@ -281,10 +290,40 @@
+
+	for (var/obj/item/weapon/stock_parts/scanning_module/S in component_parts) // scanning modules reduce power required
+		T += S.rating
+		synth_modifier = 1/T
```
This may cause a division by `0`, which you should handle by e.g. `T = max(T, 1)` on the line above.

Also, this should be done after the entire for loop has completed, once, not on every round of the iteration.

mikomyazaki replied:
The sleeper's scanner module should always have a minimum of 1 for S.rating, so T can never be zero? Unless I am missing some case? Else the machine cannot be constructed.

Are you wanting me to future-proof it against a chance someone adds a rating zero scanning module?

afterthought2 replied:
Whenever you divide by a number, and aren't 100% certain it can never be `0` (I'm not here: are you sure the things can't be destroyed/removed by anything?), it's better to be safe.

mikomyazaki replied:
I am sure this is the case - You cannot remove the scanning module without replacing with another one, or deconstructing the entire machine. But I guess it costs very little to add the check in anyway.

afterthought2 replied:
Still has the other issue, where you should move everything other than `+= S.rating` to after the for loop is complete.

afterthought2 commented on code/game/machinery/Sleeper.dm:
```diff
@@ -281,10 +290,40 @@
+/obj/machinery/sleeper/RefreshParts()
+	var/T = 0
+
+	for (var/obj/item/weapon/stock_parts/scanning_module/S in component_parts) // scanning modules reduce power required
```
Comment is misleading; this has other effects.

afterthought2 commented on code/game/machinery/Sleeper.dm:
```diff
@@ -281,10 +290,40 @@
+	for (var/obj/item/weapon/stock_parts/manipulator/M in component_parts) // manipulators unlock new drugs
+		T += M.rating
+
+	available_chemicals = base_chemicals
```
`base_chemicals.Copy()`; otherwise you'll mutate the `base_chemicals` list on the next line.

afterthought2 commented on code/game/machinery/Sleeper.dm:
```diff
@@ -281,10 +290,40 @@
+	to_chat(user, "<span class='danger'>You [emagged ? "disable" : "enable"] \the [src]'s chemical synthesis safety checks.</span>")
+
+	if (emagged)
+		available_chemicals += antag_chemicals
```
`|=` for safety is best, I think.

mikomyazaki commented:
So far as I can tell, the Travis error is not due to this PR.

afterthought2 commented on code/game/machinery/Sleeper.dm:
```diff
@@ -41,6 +47,9 @@
 		if(occupant)
 			occupant.examine(user)
+		if (emagged && usr.skill_check(SKILL_MEDICAL, SKILL_EXPERT))
+			to_chat(usr, "The sleeper chemical synthesis controls look tampered with.")
```
Still an issue; never use `usr` if it can be helped.

mikomyazaki commented:
Thanks @afterthought2 -- Should all be solved now.

afterthought2 commented:
There's still a `usr` in there.

mikomyazaki committed:
Swaps a usr to a user

Fixes PR review issues.

Makes changes based on afterthought's review

Tweaks based on feedback. Emag examine message now requires expert medicine.

Minor fix.

Makes upgrading Sleepers do something, adds emag effects.
